beacon_node/store/src/hot_cold_store.rs

```diff
@@ -977,40 +984,69 @@ impl<E: EthSpec, Hot: ItemStore<E>, Cold: ItemStore<E>> HotColdDB<E, Hot, Cold>
 
     /// Load a frozen state that lies between restore points.
     fn load_cold_intermediate_state(&self, slot: Slot) -> Result<BeaconState<E>, Error> {
+        if let Some(state) = self.state_cache.lock().get(&slot) {
+            return Ok(state.clone());
+        }
+
         // 1. Load the restore points either side of the intermediate state.
         let low_restore_point_idx = slot.as_u64() / self.config.slots_per_restore_point;
         let high_restore_point_idx = low_restore_point_idx + 1;
 
+        // Use low restore point as the base state.
+        let mut low_slot: Slot =
+            Slot::new(low_restore_point_idx * self.config.slots_per_restore_point);
+        let mut low_state: Option<BeaconState<E>> = None;
+
+        // Try to get a more recent state from the cache to avoid massive blocks replay.
+        for (s, state) in self.state_cache.lock().iter() {
+            if s.as_u64() / self.config.slots_per_restore_point == low_restore_point_idx
+                && *s < slot
+                && low_slot < *s
+            {
+                low_slot = *s;
+                low_state = Some(state.clone());
+            }
+        }
+
+        // If low_state is still None, use load_restore_point_by_index to load the state.
+        if low_state.is_none() {
+            low_state = Some(self.load_restore_point_by_index(low_restore_point_idx)?);
+        }
```

Comment:
This is just a style thing, but we usually try to avoid `.unwrap()`. In this case you can re-write this using a [`let-else`](https://doc.rust-lang.org/rust-by-example/flow_control/let_else.html).

```suggestion
        let Some(low_state) = low_state else {
            self.load_restore_point_by_index(low_restore_point_idx)?
        };
```

This creates a new variable called `low_state` that replaces the old one. Sometimes we'd call the original `opt_low_state` or something, but reusing the name is also OK.

Reply:
The code suggested here seems could not compile, it would report:
```
`else` clause of `let...else` does not diverge
expected type `!`
   found enum `types::BeaconState<E>`
try adding a diverging expression, such as `return` or `panic!(..)`
...or use `match` instead of `let...else`
```

maybe rewrite like:
```
        let low_state = match low_state {
            Some(state) => state,
            None => self.load_restore_point_by_index(low_restore_point_idx)?,
        };
```

Reply:
oh, my bad, I didn't realise `let-else` was only useable with early returns. I'm happy with the explicit match